pkg/sdk/session.go

```diff
@@ -0,0 +1,106 @@
+package sdk
+
+import (
+	"context"
+	"log"
+	"net/url"
+
+	sessiontypes "github.com/pokt-network/poktroll/x/session/types"
+	sharedtypes "github.com/pokt-network/poktroll/x/shared/types"
+)
+
+// sessionSuppliers is the structure that represents a session's end block height
+// and its matching suppliers.
+type sessionSuppliers struct {
+	// SessionEndBlockHeight is the session's end block height that helps
+	// determine if the session is still valid without looking into SupplierEndpoints slice.
+	SessionEndBlockHeight int64
+	SuppliersEndpoints    []*SupplierEndpoint
+}
+
+// SupplierEndpoint is the structure that represents a supplier's endpoint
+// augmented with the session's header and the supplier's address for easy
+// access to the needed information when sending a relay request.
+type SupplierEndpoint struct {
```

Comment:
1. I feel like `Address` and `Header` should go into `sessionSuppliers`
2. We can remove this and use the `SupplierEndpoint` struct (based on the protobuf)

Reply:
`GetCurrentSession` (now named `GetSessionSupplierEndpoints`) was initially returning the whole `Session` struct as defined in our protobufs.

The flow was that the SDK consumer gets the hydrated `Session`, performs its logic to pick a `Session.Suppliers[i].Services[serviceId].Endpoints[j].Url` then passes it to `SendRelay`.

But the fact that `SendRelay` needs more than the `Url` to send a relay (**session header**, **supplier address** and **application address**) would lead us to adopt one of with the following approaches:

1. Pass in just the `Url` and let the `SendRelay` method match it with the relevant `SupplierAddress`, `ApplicationAddress`, `ServiceId` and `RpcType` before sending the relay.
 
This approach requires the `Url` to be unique across all the `Session` and forces the `SendRelay` method to perform that search and match at each call.

2. Make the SDK consumer collect all the needed data while picking an `Url`, then pass it to `SendRelay`.

This makes the method to require at least two more arguments (`SupplierAddress` and session `Header`) in addition to `ctx`, `Url` and `request`

It puts more work and responsibility on the SDK consumer:
 * Collect the relevant info while selecting an endpoint 
 * Pass it to a `SendRelay` method which has now a long"ish" list of arguments

The approach of having a flattened list of (to be renamed) `SupplierEndpoint`s
* Generated only when `GetSession` fetches a fresh `Session` which makes it benefit from the caching mechanism;
* Returned to the SDK consumer that would only need to pick one item from it and pass it to `SendRelay`, avoiding him additional work, potential errors and further processing at the `SendRelay` level

Is the most optimal and DX friendly IMO.

We could replace `sessionSuppliers.SessionEndBlockHeight` by `sessionSuppliers.Session` and return the whole `sessionSuppliers` struct so we expose the hydrated `Session` for consumers who need it.

We should definitely rename the flattened `SupplierEndpoint` struct to avoid confusion with the original proto defined version.